From e085bc29343c634e1eb6c2b13400faf9e6b2ba4f Mon Sep 17 00:00:00 2001
From: Guoqing Jiang <Guoqing.Jiang@windriver.com>
Date: Tue, 23 Nov 2010 13:49:41 +0800
Subject: [PATCH 27/28] Mindspeed-c1k: add flash type check accordint to EXP_NAND_SEL jumper

In mindspeed-c1k board, nand and nor can't be used simultaneously,
besides use kernel config to choose flash type, also need jumper to
detect which flash type is currently support.

Signed-off-by: Guoqing Jiang <Guoqing.Jiang@windriver.com>
---
 arch/arm/mach-comcerto/board-c1kmfcn_evm.c |   35 ++++++++++++++++++++++++----
 1 files changed, 30 insertions(+), 5 deletions(-)

diff --git a/arch/arm/mach-comcerto/board-c1kmfcn_evm.c b/arch/arm/mach-comcerto/board-c1kmfcn_evm.c
index b15eb6f..11e8182 100644
--- a/arch/arm/mach-comcerto/board-c1kmfcn_evm.c
+++ b/arch/arm/mach-comcerto/board-c1kmfcn_evm.c
@@ -33,6 +33,8 @@
 #include <mach/nand.h>
 #include <mach/gpio.h>
 
+static u32 EXP_NAND_SEL;
+
 static void __init board_gpio_init(void)
 {
 #if defined(CONFIG_MTD_NAND_COMCERTO) || defined(CONFIG_MTD_NAND_COMCERTO_MODULE)
@@ -46,12 +48,13 @@ static void __init board_gpio_init(void)
 #else
 	writel(readl(COMCERTO_GPIO_PIN_SELECT_REG) | NAND_BUS, COMCERTO_GPIO_PIN_SELECT_REG);
 #endif
+	EXP_NAND_SEL = (readl(COMCERTO_GPIO_BOOTSTRAP_STATUS) & GPIO_8) >> 8 ;
 }
 
 /* --------------------------------------------------------------------
  *  NOR device
  * -------------------------------------------------------------------- */
-#if defined(CONFIG_MTD_COMCERTO_NOR)
+#if defined(CONFIG_MTD_COMCERTO_NOR) || defined(CONFIG_MTD_COMCERTO_NOR_MODULE)
 #define NORFLASH_MEMORY_PHY1 COMCERTO_EXP_CS0_BASE
 static struct resource comcerto_nor_resources[] = {
 	{
@@ -98,11 +101,22 @@ static struct platform_device comcerto_nor = {
 		.platform_data	= &comcerto_nor_data,
 	},
 };
+
+static void comcerto_init_nor(void)
+{
+	platform_device_register(&comcerto_nor);
+}
+#else
+static void comcerto_init_nor(void)
+{
+}
 #endif
 
 /* --------------------------------------------------------------------
  *  NAND device
  * -------------------------------------------------------------------- */
+#if defined(CONFIG_MTD_NAND_COMCERTO) || \
+				defined(CONFIG_MTD_NAND_COMCERTO_MODULE)
 struct mtd_partition c1kmfcn_evm_nandflash_partition[] = {
 	{
 		.name           = "u-boot",
@@ -155,6 +169,16 @@ static struct platform_device comcerto_nand = {
 	.num_resources	= ARRAY_SIZE(comcerto_nand_resources),
 };
 
+static void comcerto_init_nand(void)
+{
+	platform_device_register(&comcerto_nand);
+}
+#else
+static void comcerto_init_nand(void)
+{
+}
+#endif
+
 /* --------------------------------------------------------------------
  *  Serial interface
  * -------------------------------------------------------------------- */
@@ -485,10 +509,6 @@ static struct platform_device comcerto_usb0_device = {
 
 static struct platform_device *comcerto_devices[] __initdata = {
 	&comcerto_uart,
-	&comcerto_nand,
-#if defined(CONFIG_MTD_COMCERTO_NOR)
-	&comcerto_nor,
-#endif
 	&comcerto_fbpoolA_device,
 	&comcerto_fbpoolB_device,
 	&comcerto_mdio0_device,
@@ -535,6 +555,11 @@ static void __init platform_init(void)
 	device_init();
 	board_gpio_init();
 	platform_add_devices(comcerto_devices, ARRAY_SIZE(comcerto_devices));
+	if (EXP_NAND_SEL)
+		comcerto_init_nand();
+	else
+		comcerto_init_nor();
+
 #if defined(CONFIG_CONFIG_I2C_BOARDINFO)
 	i2c_register_board_info(0, comcerto_i2c_info,
 				ARRAY_SIZE(comcerto_i2c_info));
-- 
1.6.3.1

